File: wrappers/javascript/aries-askar-shared/src/store/Session.ts
import type { Key, SessionHandle } from '../crypto'
import type { KeyAlgs } from '../enums'

import { ariesAskar } from '../ariesAskar'
import { EntryOperation } from '../enums/EntryOperation'
import { AriesAskarError } from '../error'

import { Entry } from './Entry'
import { EntryList } from './EntryList'
import { KeyEntryList } from './KeyEntryList'

export class Session {
  private _handle?: SessionHandle
  private isTxn: boolean

  public constructor({ handle, isTxn }: { handle?: SessionHandle; isTxn: boolean }) {
    this._handle = handle
    this.isTxn = isTxn
  }

  public get isTransaction() {
    return this.isTxn
  }

  public get handle() {
    return this._handle
  }

  public async count({ category, tagFilter }: { category: string; tagFilter?: Record<string, unknown> }) {
    if (!this.handle) throw AriesAskarError.customError({ message: 'Cannot count from closed session' })
    return await ariesAskar.sessionCount({ tagFilter, category, sessionHandle: this.handle })
  }

  public async fetch({
    category,
    name,
    forUpdate = false,
    isJson = false,
  }: {
    category: string
    name: string
    forUpdate?: boolean
    isJson?: boolean
  }) {
    if (!this.handle) throw AriesAskarError.customError({ message: 'Cannot fetch from a closed session' })

    const handle = await ariesAskar.sessionFetch({ forUpdate, name, category, sessionHandle: this.handle })
    if (!handle) return undefined

    const entry = new Entry({ list: handle, position: 0 })

    return entry.toJson(isJson)
  }

  public async fetchAll({
    category,
    forUpdate = false,
    limit,
    tagFilter,
  }: {
    category: string
    tagFilter?: Record<string, unknown>
    limit?: number
    forUpdate?: boolean
  }) {
    if (!this.handle) throw AriesAskarError.customError({ message: 'Cannot fetch all from a closed session' })
    const handle = await ariesAskar.sessionFetchAll({
      forUpdate,
      limit,
      tagFilter,
      sessionHandle: this.handle,
      category,
    })
    const entryList = new EntryList({ handle })

    return entryList.toArray()
  }

  public async insert({
    category,
    name,
    expiryMs,
    tags,
    value,
  }: {
    category: string
    name: string
    value: string | Record<string, unknown>
    tags?: Record<string, unknown>
    expiryMs?: number
  }) {
    if (!this.handle) throw AriesAskarError.customError({ message: 'Cannot insert with a closed session' })
    const serializedValue = typeof value === 'string' ? value : JSON.stringify(value)

    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-call
    const encoder = new TextEncoder()

    await ariesAskar.sessionUpdate({
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment, @typescript-eslint/no-unsafe-argument, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
      value: new Uint8Array(encoder.encode(serializedValue)),
      expiryMs,
      tags,
      name,
      category,
      sessionHandle: this.handle,
      operation: EntryOperation.Insert,
    })
  }

  public async replace({
    category,
    name,
    expiryMs,
    tags,
    value,
  }: {
    category: string
    name: string
    value: string | Record<string, unknown>
    tags?: Record<string, unknown>
    expiryMs?: number
  }) {
    if (!this.handle) throw AriesAskarError.customError({ message: 'Cannot replace with a closed session' })
    const serializedValue = typeof value === 'string' ? value : JSON.stringify(value)

    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-call
    const encoder = new TextEncoder()

    await ariesAskar.sessionUpdate({
      // eslint-disable-next-line @typescript-eslint/no-unsafe-argument, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
      value: new Uint8Array(encoder.encode(serializedValue)),
      expiryMs,
      tags,
      name,
      category,
      sessionHandle: this.handle,
      operation: EntryOperation.Replace,
    })
  }

  public async remove({ category, name }: { category: string; name: string }) {
    if (!this.handle) throw AriesAskarError.customError({ message: 'Cannot remove with a closed session' })

    await ariesAskar.sessionUpdate({
      name,
      category,
      sessionHandle: this.handle,
      operation: EntryOperation.Remove,
    })
  }

  public async removeAll({ category, tagFilter }: { category: string; tagFilter?: Record<string, unknown> }) {
    if (!this.handle) throw AriesAskarError.customError({ message: 'Cannot remove all with a closed session' })

    await ariesAskar.sessionRemoveAll({
      category,
      sessionHandle: this.handle,
      tagFilter,
    })
  }

  public async insertKey({
    name,
    key,
    expiryMs,
    metadata,
    tags,
  }: {
    name: string
    key: Key
    metadata?: string
    tags?: Record<string, unknown>
    expiryMs?: number
  }) {
    if (!this.handle) throw AriesAskarError.customError({ message: 'Cannot insert a key with a closed session' })

    await ariesAskar.sessionInsertKey({
      expiryMs,
      tags,
      metadata,
      name,
      sessionHandle: this.handle,
      localKeyHandle: key.handle,
    })
  }

  public async fetchKey({ name, forUpdate = false }: { name: string; forUpdate?: boolean }) {
    if (!this.handle) throw AriesAskarError.customError({ message: 'Cannot fetch a key with a closed session' })
    const handle = await ariesAskar.sessionFetchKey({ forUpdate, name, sessionHandle: this.handle })
    const keyEntryList = new KeyEntryList({ handle })
    return keyEntryList.getEntryByIndex(0).toJson()
  }

  public async fetchAllKeys({
    forUpdate = false,
    algorithm,
    limit,
    tagFilter,
    thumbprint,
  }: {
    algorithm?: KeyAlgs
    thumbprint?: string
    tagFilter?: Record<string, unknown>
    limit?: number
    forUpdate?: boolean
  }) {
    if (!this.handle) throw AriesAskarError.customError({ message: 'Cannot fetch all keys with a closed session' })
    const handle = await ariesAskar.sessionFetchAllKeys({
      forUpdate,
      limit,
      tagFilter,
      thumbprint,
      algorithm,
      sessionHandle: this.handle,
    })

    const keyEntryList = new KeyEntryList({ handle })
    return keyEntryList.toArray()
  }

  public async updateKey({
    name,
    expiryMs,
    metadata,
    tags,
  }: {
    name: string
    metadata?: string
    tags?: Record<string, unknown>
    expiryMs?: number
  }) {
    if (!this.handle) throw AriesAskarError.customError({ message: 'Cannot update a key with a closed session' })
    await ariesAskar.sessionUpdateKey({ expiryMs, tags, metadata, name, sessionHandle: this.handle })
  }

  public async removeKey({ name }: { name: string }) {
    if (!this.handle) throw AriesAskarError.customError({ message: 'Cannot remove a key with a closed session' })
    await ariesAskar.sessionRemoveKey({ name, sessionHandle: this.handle })
  }

  /**
   * @note also closes the session
   */
  public async commit() {
    if (!this.isTxn) throw AriesAskarError.customError({ message: 'Session is not a transaction' })
    if (!this.handle) throw AriesAskarError.customError({ message: 'Cannot commit a closed session' })
    await this.handle.close(true)
    this._handle = undefined
  }

  public async rollback() {
    if (!this.isTxn) throw AriesAskarError.customError({ message: 'Session is not a transaction' })
    if (!this.handle) throw AriesAskarError.customError({ message: 'Cannot rollback a closed session' })
    await this.handle.close(false)
    this._handle = undefined
  }

  public async close() {
    if (!this.handle) throw AriesAskarError.customError({ message: 'Cannot close a closed session' })
    await this.handle.close(false)
    this._handle = undefined
  }
}

File: wrappers/kotlin/src/commonMain/kotlin/Store/Session.kt
package askar.Store

import askar.Askar
import askar.Askar.Companion.mapToJsonObject
import askar.crypto.Key
import askar.crypto.SessionHandle
import askar.enums.EntryOperation
import askar.enums.KeyAlgs
import kotlinx.cinterop.MemScope
import kotlinx.cinterop.UnsafeNumber
import kotlinx.cinterop.memScoped
import kotlinx.serialization.json.*

@OptIn(UnsafeNumber::class)
class Session(private var handle: SessionHandle?, private val isTxn: Boolean) {

    fun isTransaction(): Boolean {
        return this.isTxn
    }

    fun handle(): SessionHandle? {
        return this.handle
    }

    suspend fun count(category: String, tagFilter: String): Long {
        if (this.handle == null) throw Error("Cannot get count of closed session")
        val handle = this.handle
        var count = Askar.session.sessionCount(handle!!.handle, category, tagFilter)
        return count
    }

    suspend fun fetch(
        category: String,
        name: String,
        forUpdate: Boolean = false,
    ): EntryObject? {
        if (this.handle == null) throw Error("Cannot fetch from a close session")
        val h = Askar.session.sessionFetch(handle!!.handle, category, name, forUpdate) ?: return null
        val entry = Entry(h, 0)
        return entry.toJson()
    }

    suspend fun fetchAll(
        category: String,
        tagFilter: String = "{}",
        forUpdate: Boolean = false,
        limit: Long = -1L,
    ): ArrayList<EntryObject> {
        if (this.handle == null) throw Error("Cannot fetch from a closed session")
        val handle =
            Askar.session.fetchAll(handle!!.handle, category, tagFilter, limit, forUpdate)
                ?: return arrayListOf()
        val entryList = EntryList(handle)
        return entryList.toArray()
    }

    suspend fun insert(
        category: String,
        name: String,
        expiryMs: Long = -1,
        tags: String = "{}",
        value: String
    ): Boolean {
        if (this.handle == null) throw Error("Cannot insert into a closed session")
        val code = Askar.session.sessionUpdate(
            handle!!.handle,
            category,
            name,
            expiryMs,
            tags,
            value,
            EntryOperation.Insert
        )
        return code == 0L
    }

    suspend fun replace(
        category: String,
        name: String,
        expiryMs: Long = -1,
        tags: String = "{}",
        value: String
    ): Boolean {
        if (this.handle == null) throw Error("Cannot replace in a closed session")
        val code =
            Askar.session.sessionUpdate(
                handle!!.handle,
                category,
                name,
                expiryMs,
                tags,
                value,
                EntryOperation.Replace
            )
        return code == 0L
    }

    suspend fun remove(category: String, name: String): Boolean {
        if (this.handle == null) throw Error("Cannot remove from a closed session")
        val code =
            Askar.session.sessionUpdate(
                handle!!.handle,
                category,
                name,
                0,
                "{}",
                "",
                EntryOperation.Remove
            )
        return code == 0L
    }

    suspend fun removeAll(category: String, tagFilter: String = "{}"): Boolean {
        if (this.handle == null) throw Error("Cannot remove from a closed session")
        val code = Askar.session.sessionRemoveAll(handle!!.handle, category, tagFilter)
        return code == 0L
    }

    suspend fun insertKey(
        name: String,
        key: Key,
        expiryMs: Long = -1,
        metadata: String? = null,
        tags: String = "{}"
    ): Boolean {
        if (this.handle == null) throw Error("Cannot insert a key with a closed session")
        val code = Askar.session.sessionInsertKey(handle!!.handle, name, key, metadata, tags, expiryMs)
        return code == 0L
    }


    suspend fun fetchKey(name: String, forUpdate: Boolean = false): KeyEntryObject? {
        if (this.handle == null) throw Error("Cannot fetch key from closed session")
        val handle = Askar.session.sessionFetchKey(handle!!.handle, name, forUpdate) ?: return null
        val keyEntryList = KeyEntryList(handle)

        return keyEntryList.getEntryByIndex(0).toJson()
    }

    suspend fun fetchAllKeys(
        algorithm: KeyAlgs? = null,
        thumbprint: String? = null,
        tagFilter: String? = null,
        limit: Long = -1,
        forUpdate: Boolean = false,
    ): ArrayList<KeyEntryObject> {
        if (this.handle == null) throw Error("Cannot fetch keys from a closed session")
        val handle = Askar.session.sessionFetchAllKeys(
            handle!!.handle,
            algorithm,
            thumbprint,
            tagFilter,
            limit,
            forUpdate,
        ) ?: return ArrayList()
        val keyEntryList = KeyEntryList(handle)
        return keyEntryList.toArray()
    }

    suspend fun updateKey(
        name: String,
        metadata: String? = null,
        tags: String = "{}",
        expiryMs: Long = -1
    ): Boolean {
        if (this.handle == null) throw Error("Cannot update key from a closed session")
        val code = Askar.session.sessionUpdateKey(handle!!.handle, name, metadata, tags, expiryMs)
        return code == 0L
    }

    suspend fun removeKey(name: String): Boolean {
        if (this.handle == null) throw Error("Cannot remove key from a closed session")
        val code = Askar.session.sessionRemoveKey(handle!!.handle, name)
        return code == 0L
    }

    /**
     * @Note also closes the session
     */
    suspend fun commit(): Boolean {
        if (!this.isTxn) throw Error("Session is not a transaction")
        if (this.handle == null) throw Error("Cannot commit a closed session")
        val code = handle!!.close(true)
        this.handle = null
        return code == 0L
    }

    suspend fun rollback(): Boolean {
        if (!this.isTxn) throw Error("Session is not a transaction")
        if (this.handle == null) throw Error("Cannot rollback a closed session")
        val code = handle!!.close(false)
        this.handle = null
        return code == 0L
    }

    suspend fun close(): Boolean {
        if (this.handle == null) throw Error("Cannot close a closed session")
        val code = handle!!.close(false)
        this.handle = null
        return code == 0L
    }


}